### ISSUE TITLE
Metric support

### ISSUE THREAD

**Issue by t11omas** [zarusz/SlimMessageBus#333]
Are their any plans to introduce metrics, for example the duration of a consumer?  I can write an interceptor to do this currently, but I don't thinks its an accurate value as the message has already had consumption logic performed on it, i,e desterilized at this point.

So I think either a hook in point to when the message is first received is required, or build the metrics into SMB

**Comment by zarusz** [comment]
hello,

So I was thinking about adding these at some point via the OTEL.

**Comment by t11omas** [comment]
I have done that via the generic interceptor, but the issue is that by the time it hits the interceptor, some processing has already occurred.  I.e, its already been desterilized which means that time wont be included.

I will take a look at https://github.com/zarusz/SlimMessageBus/issues/149 :)

**Comment by t11omas** [comment]
Just copying this question over to this ticket:

Is there any way to see/monitor the count of messages on the in memory bus? (might be another example of a good metric to log)

We have a process that reads a lot of data and places it onto an in memory bus to be processed. It would be nice to be able to see/monitor that in memory bus

**Comment by zarusz** [comment]
Yes, that's another good metric that should be logged.

For now you could either turn on more [verbose logging](https://github.com/zarusz/SlimMessageBus/blob/master/docs/intro.md#logging) and then count the lifecycle log events that are indicative of counts (not great, but possible).

Add an generic interceptor as explained above and increment the counter for your metric platform, but you have that already.
For in memory by default messages are not serialized/deserialized so it wont include that overhead (which is what you're looking for).

Either way, we need to capture metrics (new feature). I belive we could use OTEL as the abstraction.
Would OTEL work for you? What logging/monitoring platform do you use?

**Comment by t11omas** [comment]
We are using a Hybrid bus, so both in memory and external.  I understand for the in memory bus the messages don't go via serialization, but they do for the external buses.

OTEL would work for us, but I am wondering, just from a design point of via, could it somehow be abstracted like what you have done for the transports and serialization.

For example, and don't get caught up on the names or signatures I use here, they are just used to illustrate the point.

```
public interface IConsumerObserver 
{
   void ConsumeStarted();

   void MessageErrored();
 
   void  ConsumeFinished();

}
```

```
public interface IInMemoryBusObserver
{
   void MessageAdded();

   void MessagedRemoved();

}
```

That way you can added in OTELMetric or extend it to support whatever platform is required

**Comment by zarusz** [comment]
Yes, that what I was thinking too. OTEL isn't a clean abstraction (we have to drag in libraries into the SlimMessageBus.Host) in itself so that would address my concern. However, OTEL becomes a standard these days that many platforms use nowadays, and I believe users of SMB would expect to just connect it to Prometheus/Grafana/Azure App Insights and it simply working.

Maybe the path would be SMB > SMBs abtraction  (the interface above as suggested) > OTEL plugin (optional).

I will think about it. Again, please voice your needs and opinions. Feedback is much appreciated.

**Comment by zarusz** [comment]
Reading some more about [metrics instrumentation](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/metrics-instrumentation) it seems that nowadays the best and modern way to add metrics to .net apps is via [System.Diagnostics.DiagnosticSource](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource). It is a good and performant abstraction with OTEL in mind and tools like Prometheus and Grafana can collect the metrics.

I will give it a try.

**Comment by t11omas** [comment]
If it helps, this is what i was playing around with using the interceptor:

```
public class MetricsConsumerInterceptor<TMessage> : IConsumerInterceptor<TMessage>
{
    readonly Histogram<double> consumeDuration;
    readonly Counter<long> consumeTotal;
    readonly Counter<long> consumeFaultTotal;
    readonly Counter<long> consumerInProgress;
    readonly Counter<long> consumeRetryTotal;

    public MetricsConsumerInterceptor(IMeterFactory meterFactory)
    {
        var meter = meterFactory.Create("Cdms");
        consumeTotal = meter.CreateCounter<long>("messaging.cdms.consume", "ea", "Number of messages consumed");
        consumeFaultTotal = meter.CreateCounter<long>("messaging.cdms.consume.errors", "ea",
            "Number of message consume faults");
        consumerInProgress = meter.CreateCounter<long>("messaging.cdms.consume.active", "ea",
            "Number of consumers in progress");
        consumeDuration = meter.CreateHistogram<double>("messaging.cdms.consume.duration", "ms",
            "Elapsed time spent consuming a message, in millis");
        consumeRetryTotal =
            meter.CreateCounter<long>("messaging.cdms.consume.retries", "ea", "Number of message consume retries");
    }

    public async Task<object> OnHandle(TMessage message, Func<Task<object>> next, IConsumerContext context)
    {
        var timer = Stopwatch.StartNew();
        var tagList = new TagList
        {
            { "messaging.cdms.service", Process.GetCurrentProcess().ProcessName },
            { "messaging.cdms.destination", context.Path },
            {
                "messaging.cdms.message_type",
                ObservabilityUtils.FormatTypeName(new StringBuilder(), typeof(TMessage))
            },
            { "messaging.cdms.consumer_type", context.Consumer.GetType().Name }
        };

        try
        {
            consumeTotal.Add(1, tagList);
            consumerInProgress.Add(1, tagList);
            if (context.Properties.TryGetValue(MessageBusHeaders.RetryCount, out var value))
            {
                tagList.Add("messaging.cdms.retry_attempt", (int)value);
                consumeRetryTotal.Add(1, tagList);
            }

            return await next();
        }
        catch (Exception exception)
        {
            tagList.Add("messaging.cdms.exception_type", exception.GetType().Name);
            consumeFaultTotal.Add(1, tagList);
            throw;
        }
        finally
        {
            consumerInProgress.Add(-1, tagList);
            consumeDuration.Record(timer.ElapsedMilliseconds, tagList);
        }
    }
}
```

.Net abstraction does come with an OTEL Exporter.  We might need to use EMF instead though.  I am just investigation on whether we can export the .net abstraction to EMF

**Comment by zarusz** [comment]
Thanks for sharing. The interceptor example you've provided was one of the ways I was looking to implement this, however, exploring if adding this internally (outside of the interceptors pipeline) to make this more efficient.

Couple of questions, so let me know your thoughts here:

1. I see you've used one metric consume and the path is just a tag on it. I was wondering if doing the metric per path/topic would be more desirable (each path/topic would get its own total count metric).  Applying the path should give a way to see all the consumers metic as one or segregate them by topic. So perhaps that's the most flexible, but please comment here.

2. What is the reason you'd like to see metrics captured at the native message (e.g. ServiceBusMessage)? The challenge here is that little is known about the message type at that stage and the types per transports vary making it troublesome to write one unified metric. On the other hand the overhead for serialization (or other SMB ceremonies) should be negligible. Let me know.

3. I am thinking about giving a away to configure the metric names and prefix, as well as to be able to augment the tags collected (to collect less or more).

4. Do we want producer metrics too?
